pkg/dockerregistry/server/projectcache.go

```diff
@@ -0,0 +1,103 @@
+// Used to store a list of items per namespace.
+type projectObjectListStore interface {
+	Add(namespace string, obj runtime.Object) error
+	Get(namespace string) (obj runtime.Object, exists bool, err error)
+}
+
+// projectObjectListCache implements projectObjectListStore.
+type projectObjectListCache struct {
+	cache.Store
+}
+
+var _ projectObjectListStore = &projectObjectListCache{}
+
+// newProjectResourceQuotaListCache creates a cache to hold ResourceQuotaList objects
+// thal will expire with the given ttl.
+func newProjectResourceQuotaListCache(ttl time.Duration) projectObjectListStore {
+	return &projectObjectListCache{
+		Store: cache.NewTTLStore(metaProjectResourceQuotaListKeyFunc, ttl),
+	}
+}
+
+// newProjectLimitRangeList creates a cache to hold LimitRangeList objects
+// that will expire with the given ttl.
+func newProjectLimitRangeListCache(ttl time.Duration) projectObjectListStore {
+	return &projectObjectListCache{
+		Store: cache.NewTTLStore(metaProjectLimitRangeListKeyFunc, ttl),
+	}
+}
+
+// Add stores given list object under the given namespace. Any prior object under this
+// key will be replaced.
+func (c *projectObjectListCache) Add(namespace string, obj runtime.Object) error {
+	if namespace == "" {
+		return fmt.Errorf("namespace cannot be empty")
+	}
+	no := &namespacedObject{
+		namespace: namespace,
+		object:    obj,
+	}
+	return c.Store.Add(no)
+}
+
+// Get retrieves a cached list object if present and not expired.
+func (c *projectObjectListCache) Get(namespace string) (obj runtime.Object, exists bool, err error) {
+	o, exists, err := c.Store.GetByKey(namespace)
+	if exists {
+		no := o.(*namespacedObject)
+		obj = no.object
+	}
+	return
+}
+
+// namespacedObject is a container associating an object with a namespace.
+type namespacedObject struct {
+	namespace string
+	object    runtime.Object
+}
+
+// metaProjectResourceQuotaListKeyFunc returns a key for given namespaced object and
+// ensures it holds ResourceQuotaList. The key is object's namespace.
+func metaProjectResourceQuotaListKeyFunc(obj interface{}) (string, error) {
+	if key, ok := obj.(cache.ExplicitKey); ok {
+		return string(key), nil
+	}
+	no, ok := obj.(*namespacedObject)
+	if !ok {
+		return "", fmt.Errorf("object %T is not a namespaced object", obj)
+	}
+	_, ok = no.object.(*kapi.ResourceQuotaList)
+	if !ok {
+		return "", fmt.Errorf("namespaced object %T is not a resource quota list", obj)
+	}
+	return no.namespace, nil
+}
+
+// metaProjectLimitRangeListKeyFunc returns a key for given namespaced object and
+// ensures it holds LimitRangeList. The key is object's namespace.
+func metaProjectLimitRangeListKeyFunc(obj interface{}) (string, error) {
+	if key, ok := obj.(cache.ExplicitKey); ok {
+		return string(key), nil
+	}
+	no, ok := obj.(*namespacedObject)
+	if !ok {
+		return "", fmt.Errorf("object %T is not a namespaced object", obj)
+	}
+	_, ok = no.object.(*kapi.LimitRangeList)
```

Comment:
Why this check?  Why not collapse to a common method and leave casting at the calling layer where you're still having to do it on gets?